llvm/lib/IR/Mangler.cpp

```diff
@@ -302,14 +302,23 @@ std::optional<std::string> llvm::getArm64ECMangledFunctionName(StringRef Name) {
   // Insert the ARM64EC "$$h" tag after the mangled function name.
   if (Name.contains("$$h"))
     return std::nullopt;
-  size_t InsertIdx = Name.find("@@");
-  size_t ThreeAtSignsIdx = Name.find("@@@");
-  if (InsertIdx != std::string::npos && InsertIdx != ThreeAtSignsIdx) {
+
+  // The last 4 characters of the symbol type may contain a `@@` if the symbol
+  // is returning a qualified type. We don't want to insert `$$h` at that point.
+  auto TrimmedName = Name.drop_back(4);
+
+  // The last `@@` is the separation between the qualified name of the symbol
+  // and its type, which is where we want to insert `$$h`.
```

Comment:
MSVC generates the mangled name in the frontend, and that's certainly the most robust choice.

Unfortunately, LLD and import library generation have both taken a dependency on this function so we'd have to detangle that first.

The other option is to use a name demangler to parse the mangled name instead of this adhoc from-the-back parsing.

Reply:
Right.

I'd guess parsing forward from the front is easier than parsing backwards, because that's the way mangling naturally works.  And I'm not sure there are any good shortcuts here.  Not sure how much code a "mini-demangler" that knows enough to parse the name of the function without actually printing the demangled name would be, vs. just pulling in the real demangler.

Reply:
Turns out using the real demangler was fairly easy. Also added your new test case.